### PR TITLE
Instance based subscription cache for cursor service

### PR DESCRIPTION
**Pull request by adyach**
# One-line summary
Instance based subscription cache for cursor service, it is not distributed caching solution rather simple improvement to tolerate database failures and unnecessary calls to database.
The lazing caching is used, invalidation happens only if deletion of subscription happens on the same instance or 5 minutes interval after last access by key.

### CODE REVIEW

File: src/main/java/org/zalando/nakadi/service/SubscriptionCache.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+import org.zalando.nakadi.domain.Subscription;
+import org.zalando.nakadi.exceptions.runtime.NoSuchSubscriptionException;
+import org.zalando.nakadi.exceptions.runtime.ServiceTemporarilyUnavailableException;
+import org.zalando.nakadi.repository.db.SubscriptionDbRepository;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * The caching works only inside the instance of the application, it does not sync cahce across the instances, that's
```
Comment (rcillo) [inline src/main/java/org/zalando/nakadi/service/SubscriptionCache.java]: `not sync cahce/not sync cache`

File: src/main/java/org/zalando/nakadi/service/SubscriptionCache.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+     * @throws ServiceTemporarilyUnavailableException
+     */
+    public Subscription getSubscription(final String subscriptionId)
+            throws NoSuchSubscriptionException, ServiceTemporarilyUnavailableException {
+        try {
+            return subscriptionsCache.get(subscriptionId);
+        } catch (final UncheckedExecutionException e) {
+            final Throwable cause = e.getCause();
+            if (cause instanceof NoSuchSubscriptionException) {
+                throw new NoSuchSubscriptionException(e.getCause().getMessage());
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/SubscriptionCache.java]: Why not to throw cause ?

File: src/main/java/org/zalando/nakadi/service/SubscriptionCache.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            }
+            throw new ServiceTemporarilyUnavailableException(e.getCause().getMessage());
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/SubscriptionCache.java]: Why not to add cause?

File: src/main/java/org/zalando/nakadi/service/SubscriptionCache.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        } catch (final ExecutionException e) {
+            throw new ServiceTemporarilyUnavailableException(e.getCause().getMessage());
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/SubscriptionCache.java]: The same question

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionService.java [hunk 5]
```diff
@@ -145,6 +148,7 @@ public Subscription updateSubscription(final String subscriptionId, final Subscr
         subscriptionValidationService.validateSubscriptionChange(old, newValue);
         old.mergeFrom(newValue);
         subscriptionRepository.updateSubscription(old);
+        subscriptionCache.invalidateSubscription(subscriptionId);
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionService.java]: I believe this call is useless when we have 150 instances

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionService.java [hunk 6]
```diff
@@ -213,6 +217,7 @@ public void deleteSubscription(final String subscriptionId)
         nakadiKpiPublisher.publish(subLogEventType, () -> new JSONObject()
                 .put("subscription_id", subscriptionId)
                 .put("status", "deleted"));
+        subscriptionCache.invalidateSubscription(subscriptionId);
```
Comment (antban) [inline src/main/java/org/zalando/nakadi/service/subscription/SubscriptionService.java]: I believe this call is useless when we have 150 instances

**Comment by rcillo** [comment]
👍 this is a great addition of value. I would even say that it's not necessary to invalidate on the instance. Rather, add a sentence to the API saying that subscription deletion is eventually consistent and cached.

Do you know what's the behaviour of a commit made to a subscription that has been deleted but the cache did not expire yet? Does it fail with 500?

File: src/main/java/org/zalando/nakadi/service/subscription/SubscriptionService.java (hunk 6 above)
Comment (adyach) [reply src/main/java/org/zalando/nakadi/service/subscription/SubscriptionService.java]: but what if you have 1 instance?

**Comment by rcillo** [comment]
@andrey if it would not be too much work to gracefully handle this 500, that would be ideal.

**Comment by rcillo** [comment]
@adyach I'm even fine with adding a ticket for that, if it's too much work, I would approve this PR.

**Comment by adyach** [comment]
@rcillo it is quite hard to avoid it, because cache is not distributed. I have left cache only for commits, so there will be no 500s, but I will have 422 due to invalid session. In order to avoid 422, I will have to add additional call to zk, when session is invalid, I decided not to do it at this point

**Comment by rcillo** [comment]
👍

**Comment by adyach** [comment]
👍

**Comment by adyach** [comment]
👍

**Comment by rcillo** [comment]
👍

**Comment by adyach** [comment]
👍

**Comment by rcillo** [comment]
👍

**Pull request merged**